core/src/main/java/org/apache/calcite/adapter/enumerable/RexToLixTranslator.java

```diff
@@ -1061,10 +1061,22 @@ public static Expression convert(Expression operand, Type fromType,
           Expressions.convert_(operand, toBox.primitiveClass),
           toBox);
     } else if (fromType == java.sql.Date.class) {
-      if (toBox == Primitive.INT) {
+      if (toType == int.class) {
         return Expressions.call(BuiltInMethod.DATE_TO_INT.method, operand);
-      } else {
-        return Expressions.convert_(operand, toType);
+      } else if (toType == Integer.class) {
+        return Expressions.call(BuiltInMethod.DATE_TO_INT_OPTIONAL.method, operand);
```

Comment:
Could here be merged to use `BuiltInMethod.DATE_TO_INT_OPTIONAL` for both int and Integer?

Reply:
They can't be merged together. During codegen, Calcite will check methods through reflection to compare types.

Reply:
I am surprised. I thought `Integer` is only used in place of `int` if there are null values; the code generator checks for null values, and if argument is null and the method is strong, does not even call the method. Thus we usually only require methods with primitive argument and return types.

Reply:
Yes, the code generator prefers primitive even it is unsafe to unbox sometimes. Acutally, `DATE` to `Integer`, `TIME `to `Integer` and `TIMESTAMP` to `LONG` casting will never been called in current codegen, because it has been unboxed before calling the method.
There are two reasons to make these changes in this PR.
(1)`RexToLixTranslator.convert()` is a static function that provides general type cast support. We should ensure its correctness.
(2)Current codegen defines `AlwaysNull` exception for unusual situations, in which it returns `null` directly. This mechanism buries a lot of things and also brings many tricks to handle. It is nececessary to fix the underlying problems.

Reply:
I also have this concern, the is null check should be handled by the code generator right before entering this `convert` method, thus, we can make the logic clean. So we do not need method like `BuiltInMethod.DATE_TO_INT_OPTIONAL`

Reply:
@danny0405 I incline to fix them.
On one hand, `RexToLixTranslator.convert()` is a public and static interface. The null check policy is only guaranteed in current codegen framework. We cannot prevent userd or developers to call it. On the other hand, with the evolution of codegen, more problem will exposed. For example, another pattern in [CALCITE-3520](https://issues.apache.org/jira/browse/CALCITE-3520). I add a new test file for these underlying issues.